tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/LicenseChecker.java

```diff
@@ -0,0 +1,320 @@
+
+		// check that all required NOTICE files exists
+		severeIssueCount += ensureRequiredNoticeFiles(modulesWithShadedDependencies, noticeFiles);
+
+		// check each NOTICE file
+		for (File noticeFile: noticeFiles) {
+			severeIssueCount += checkNoticeFile(modulesWithShadedDependencies, noticeFile);
+		}
+
+		// find modules included in flink-dist
+
+		return severeIssueCount;
+	}
+
+	private int ensureRequiredNoticeFiles(Multimap<String, IncludedDependency> modulesWithShadedDependencies, List<File> noticeFiles) {
+		int severeIssueCount = 0;
+		Set<String> shadingModules = new HashSet<>(modulesWithShadedDependencies.keys());
+		shadingModules.removeAll(noticeFiles.stream().map(LicenseChecker::getModuleFromNoticeFile).collect(Collectors.toList()));
+		for (String moduleWithoutNotice : shadingModules) {
+			if (!MODULES_SKIPPING_DEPLOYMENT.contains(moduleWithoutNotice)) {
+				LOG.warn("Module {} is missing a NOTICE file. It has shaded dependencies: {}", moduleWithoutNotice, modulesWithShadedDependencies.get(moduleWithoutNotice));
+				severeIssueCount++;
+			}
+		}
+		return severeIssueCount;
+	}
+
+	private static String getModuleFromNoticeFile(File noticeFile) {
+		File moduleFile = noticeFile.getParentFile() // META-INF
+			.getParentFile() // resources
+			.getParentFile() // main
+			.getParentFile() // src
+			.getParentFile(); // <-- module name
+		return moduleFile.getName();
+	}
+
+	private int checkNoticeFile(Multimap<String, IncludedDependency> modulesWithShadedDependencies, File noticeFile) throws IOException {
+		int severeIssueCount = 0;
+		String moduleName = getModuleFromNoticeFile(noticeFile);
+
+		// 1st line contains module name
+		String noticeContents = readFile(noticeFile.toPath());
+		if (!noticeContents.startsWith(moduleName)) {
+			String firstLine = noticeContents.substring(0, noticeContents.indexOf('\n'));
+			LOG.warn("Expected first file of notice file to start with module name. moduleName={}, firstLine={}", moduleName, firstLine);
+		}
+
+		// collect all declared dependencies from NOTICE file
+		Set<IncludedDependency> declaredDependencies = new HashSet<>();
+		try (BufferedReader br = new BufferedReader(new StringReader(noticeContents))) {
+			String line;
+			while ((line = br.readLine()) != null) {
+				Matcher noticeDependencyMatcher = NOTICE_DEPENDENCY_PATTERN.matcher(line);
+				if (noticeDependencyMatcher.find()) {
+					String groupId = noticeDependencyMatcher.group(1);
+					String artifactId = noticeDependencyMatcher.group(2);
+					String version = noticeDependencyMatcher.group(3);
+					if (groupId == null && artifactId == null && version == null) { // "bundles" case
+						groupId = noticeDependencyMatcher.group(5);
+						artifactId = noticeDependencyMatcher.group(6);
+						version = noticeDependencyMatcher.group(7);
+					}
+					IncludedDependency toAdd = IncludedDependency.create(groupId, artifactId, version);
+					if (!declaredDependencies.add(toAdd)) {
```

Comment:
this should be considered a severe issue, because it could also happen if 2 artifacts with the same groupId/artifactId/version but different types are bundled.
If this ever happens, we want this to fail _somewhere_.